native/dgebrd.go

```diff
@@ -0,0 +1,92 @@
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dgebrd reduces a general m×n matrix A to upper or lower bidiagonal form B by
+// an orthogonal transformation: Q^T * A * P = B.
+func (impl Implementation) Dgebrd(m, n int, a []float64, lda int, d, e, tauQ, tauP, work []float64, lwork int) {
+	checkMatrix(m, n, a, lda)
+	minmn := min(m, n)
+	if len(d) < minmn {
+		panic(badD)
+	}
+	if len(e) < minmn-1 {
+		panic(badE)
+	}
+	if len(tauQ) < minmn {
+		panic(badTauQ)
+	}
+	if len(tauP) < minmn {
+		panic(badTauP)
+	}
+	// Calculate optimal work.
+	nb := impl.Ilaenv(1, "DGEBRD", " ", m, n, -1, -1)
+	if lwork == -1 {
+		lworkOpt := (m + n) * nb
+		work[0] = float64(lworkOpt)
+		return
+	}
+	ws := max(m, n)
+	if lwork < ws {
+		panic(badWork)
+	}
+	if len(work) < lwork {
+		panic(badWork)
+	}
+	var nx int
+	if nb > 1 && nb < minmn {
+		nx = max(nb, impl.Ilaenv(3, "DGEBRD", " ", m, n, -1, -1))
+		if nx < minmn {
+			ws = (m + n) * nb
+			if lwork < ws {
+				nbmin := impl.Ilaenv(2, "DGEBRD", " ", m, n, -1, -1)
+				if lwork >= (m+n)*nbmin {
+					nb = lwork / (m + n)
+				} else {
+					nb = 1
+					nx = minmn
+				}
+			}
+		}
+	} else {
+		nx = minmn
+	}
+	bi := blas64.Implementation()
+	ldworkx := nb
+	ldworky := nb
+	// Netlib lapack has minmn - nx, but this makes the last nx rows (which by
+	// default is large) be unblocked. As written here, the blocking is more
+	// consistent.
+	var i int
+	for i = 0; i < minmn-nb; i += nb {
+		// Reduce rows and columns i:i+nb to bidiagonal form and return
+		// the matrices X and Y which are needed to update the unreduced
+		// part of the matrix.
+		x := work[:m*nb]
```

Comment:
Yea. X is stored in the first m rows, and y in the next rows. Updated comment and code for clarity.